### DISCUSSION TITLE
DataPortal.Fetch not finding my Fetch method when using [Inject]

### DISCUSSION THREAD

**Discussion by oskarjf** [MarimerLLC/csla#3517] [Questions] [answered]
**I'm using CSLA 5.5.3 and .Net 7.**

I have a Fetch method that is defined like this:
```
[Fetch]
private void Fetch(MyCriteria criteria, [Inject] IMyService myService)
{ }
```

I then call this method with
`DataPortal.Fetch<SomeObj>(new MyCriteria(aId, bId));`

For some reason CSLA can't find the Fetch method and throws an exception :
```
Csla.DataPortalException
  HResult=0x80131500
  Message=DataPortal.Fetch failed (service)
  Source=Csla
  StackTrace:
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
```

If I remove the [Inject] parameter the DataPortal can find my method just fine.

Should I be injecting dependencies in another way? How can I fix this?

**Comment by oskarjf** [comment]
One note is that I have multiple Fetch methods with different criteria parameters.

**Reply by oskarjf** [reply]
The strange thing though is that I've now created unit tests for these failing methods and they work.  CSLA seems to be able to find the correct Fetch method and inject dependent services.

The error occurs from another service project that is using this library (within the same solution).

**Reply by rockfordlhotka** [reply]
Can you show the calling code?

Or perhaps the required service isn't registered in the host process that's running the code?

**Reply by rockfordlhotka** [reply]
You aren't getting a parameter mis-match exception - something about unable to find a matching method. Instead you seem to be getting an exception about the service - so I'm now thinking that the data portal (via DI) is unable to create an instance of the required service based on the `Inject` attribute.

**Reply by oskarjf** [reply]
The injected service is injected in constructors with Microsoft DI in many places without any issues.

The class that contains the Fetch methods:
`public class Groups : BusinessListBase<Groups, Group>`

The method that gets called externally
```
public static Groups GetGroupListForUser(int userId, bool onlyActive = false)
{
    return DataPortal.Fetch<Groups>(new CriteriaUserId(userId, onlyActive));
}
```
The DataPortal Fetch method
```
[Fetch]
protected void DataPortal_Fetch(CriteriaUserId criteria, [Inject]ICommonService commonService)
```

As I said I've got a unit test which calls the exposed method in the same manner as the one failing but I don't get any exception in the unit test.
Unit test call
`Groups.GetGroupListForUser(userId, onlyActive);`

**Comment by jhuerta3** [comment]
Maybe the following needs to be on Server project  on  Programs.cs  

builder.Services.AddTransient(typeof(DataAccess.IXXXXDal), typeof(DataAccess.EF.XXXXEFDal));

**Comment by oskarjf** [comment, marked as answer]
This was solved by adding this line to the Program.cs
`ApplicationContext.DefaultServiceProvider = services.BuildServiceProvider();`

For some reason CSLA wasn't picking up the service provider.

**Reply by rockfordlhotka** [reply]
fwiw, we've completely reworked how `ApplicationContext` finds and uses the service provider in later versions of CSLA.

CSLA 5 was the first real attempt to use DI and make it optional. Turns out that was a bad strategy, and so current versions of CSLA _require_ DI and avoid a whole host of issues from those early v5 days.
